### PR TITLE
0.3.0 - Validation outcome

### PR DESCRIPTION
**Pull request by CalebCourier**
This PR contains a breaking change that consolidates all returnables from `__call__` and `parse` into a single class `ValidationOutcome`.  It also adds a property onto this class for easy determination of whether validation succeeded or failed.

Addresses: https://github.com/ShreyaR/guardrails/issues/283

### CODE REVIEW

**Comment by CalebCourier** [comment]
The tests that are failing are failing on main.  Will put in a different PR to fix those separately.

[Edit] PR for test fixes on main: https://github.com/guardrails-ai/guardrails/pull/14

**Comment by CalebCourier** [comment]
@irgolic I switched from overloads to generics like we talked about and I think that makes it a little cleaner.  It also has the semi-desirable side-effect of turning Guard into a generic as well that we type during the class methods.  In order to make the typing a little more strict I also made the validated_output distinct from the final reask for cases where we run out of retries.  The one caveat here is that if we instantiate the Guard from a rail spec we still end up with a union type since we don't know the output_schema up front and trying to infer it at runtime just results in the intellisense defaulting to Any.

**Comment by CalebCourier** [comment]
@zsimjee We need to discuss timing for this one since it's a breaking change.  We released 0.2.0 three weeks ago, so according to our the release cycle we defined in our versioning guide, we could release this as 0.3.0 as early as next week (the 1 month mark).

~Unknowns that should affect our decision:~

~- Can we still support destructuring the result with this new class?  If so then this is only a breaking change for parse.~
We can support destructuring by defining `__iter__` and `__getitem__` on the class.  This way a user executing the Guard via `__call__` will not have to change their code.

File: guardrails/classes/validation_outcome.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from typing import Dict, Generic, Optional, TypeVar
+
+from pydantic import Field
+
+from guardrails.utils.logs_utils import ArbitraryModel, GuardHistory
+from guardrails.utils.reask_utils import ReAsk
+
+T = TypeVar("T", str, Dict, None)
```
Comment (irgolic) [inline guardrails/classes/validation_outcome.py]: When would T be `None`?

Change looks reasonable otherwise :)

Comment (CalebCourier) [reply guardrails/classes/validation_outcome.py]: With this latest iteration `T` would be `None` when validation fails and we have run out of reasks; aka when the new `reask` property is not `None`.

Comment (irgolic) [reply guardrails/classes/validation_outcome.py]: `validated_output` is already optional, I think it makes more sense for `T` to be deterministic, invariant to anything except how you construct the guard

Comment (CalebCourier) [reply guardrails/classes/validation_outcome.py]: > validated_output is already optional

Good point, I didn't catch that as I was separating `reask` from `validated_output`

**Pull request closed without merging**